docs/tutorial/using-request-directly.md

```diff
@@ -53,3 +53,63 @@ Then declare a *path operation function* parameter with the type being the `Requ
 ## `Request` documentation
 
 You can read more details about the <a href="https://www.starlette.io/requests/" target="_blank">`Request` object in the official Starlette documentation site</a>.
+
+## Using a custom `Request` class
+
+!!! danger
+    This is an "advanced" feature.
+    
+    If you are just starting with **FastAPI** you might want to skip this section.
+
+In some cases, it may be desirable to override the logic used by the `Request` class.
+
+In particular, this may be a good alternative to middleware if you want to read or manipulate the request body before it is processed by your application.
+
+Some use cases include:
+
+* converting non-JSON request bodies to JSON (e.g., `msgpack`)
+* decompressing gzip-compressed request bodies
+* automatically logging all request bodies
+* accessing the request body in an exception handler
+
+### Handling custom request body encodings
+Let's see how to make use of a custom `Request` subclass to decompress gzip requests.
+
+First, we create a `GzipRequest` class, which will override `Request.body` to decompress the body in the presence of an appropriate header: 
+
+```Python hl_lines="10"
+{!./src/using_request_directly/tutorial002.py!}
+```
+
+Next, we create a custom subclass of `fastapi.routing.APIRoute` that will make use of the `GzipRequest`:
+
+```Python hl_lines="20"
+{!./src/using_request_directly/tutorial002.py!}
+```
+
+!!! info
+    The `get_app` method uses all of the information stored on the `APIRoute` to *build* a callable that accepts a `Request`, and returns a `Response`.
+
+    The name doesn't refer to your *whole app*, it just reflects the additional pre- and post-processing logic beyond the endpoint call.
```

Comment:
Not sure if I'm describing this correctly 😅 just wanted to flag that.

Also I'm not sure if the new `APIRoute.get_app` should be renamed. Since it just calls the `get_app` function that seemed like a good name to me though.

Reply:
Good point!

The name `get_app` is probably confusing for others.

The original name was based on how it was used in Starlette, which in turn was based on the ASGI spec (technically, a request handler is an ASGI app itself, that could stand by itself).

---

But it could clearly be confusing. I think we should rename it.

How about something like:

*  `request_function`
* `get_request_function`
* `request_handler`
* `get_request_handler`
* `get_handler`

What do you think?

Also, I think we should name the stand-alone function one way and the class method something different, to avoid the editor linking the function from inside the method to the method itself (instead of the function that lives outside).

Reply:
I'm in favor of `get_request_handler` for the function and `get_route_handler` for the method. (As discussed on gitter.)